bundles/org.eclipse.sirius.emfjson/src/main/java/org/eclipse/sirius/emfjson/utils/GsonEObjectSerializer.java

```diff
@@ -1040,22 +1043,45 @@ private JsonElement serializeEDataType(EObject eObject, EAttribute eAttribute) {
             if (value instanceof Collection) {
                 Collection<?> collection = (Collection<?>) value;
                 for (Object object : collection) {
-                    jsonArray.add(new JsonPrimitive(eFactoryInstance.convertToString(eAttribute.getEAttributeType(), object)));
+                    if (object == null) {
+                        jsonArray.add(JsonNull.INSTANCE);
+                    } else if (this.shouldEDataTypeBeSerializedInJson(eAttribute.getEAttributeType())) {
+                        jsonArray.add(new Gson().toJsonTree(object));
+                    } else {
+                        jsonArray.add(new JsonPrimitive(eFactoryInstance.convertToString(eAttribute.getEAttributeType(), object)));
+                    }
                 }
             }
             jsonElement = jsonArray;
         } else {
             String stringValue = eFactoryInstance.convertToString(eAttribute.getEAttributeType(), value);
-
             if (stringValue == null) {
-                stringValue = ""; //$NON-NLS-1$
+                jsonElement = JsonNull.INSTANCE;
+            } else if (this.shouldEDataTypeBeSerializedInJson(eAttribute.getEAttributeType())) {
+                jsonElement = new Gson().toJsonTree(value);
+            } else {
+                jsonElement = new JsonPrimitive(stringValue);
             }
```

Comment:
I think that `shouldEDataTypeBeSerializedInJson()` should be called first, like:

```java
 if (this.shouldEDataTypeBeSerializedInJson(eAttribute.getEAttributeType())) {
      jsonElement = new Gson().toJsonTree(value);
  } else {
      String stringValue = eFactoryInstance.convertToString(eAttribute.getEAttributeType(), value);
      if (stringValue == null) {
          jsonElement = JsonNull.INSTANCE;
      } else {
          jsonElement = new JsonPrimitive(stringValue);
      }
  }
```

It's the same beahviour of the `eAttribute.isMany()` case above.